csrc/ir/nodes.cpp

```diff
@@ -4355,7 +4356,14 @@ std::vector<PolymorphicValue> MatmulOp::evaluate(
     const std::vector<PolymorphicValue>& inputs) const {
   const auto a = inputs.at(0).as<at::Tensor>();
   const auto b = inputs.at(1).as<at::Tensor>();
-  return {at::matmul(a, b)};
+  auto matmul_out = at::matmul(a, b);
+  if (out()->hasAllocation()) {
+    auto matmul_sizes = matmul_out.sizes().vec();
+    auto strides = computeStrides(out(), matmul_sizes);
+    matmul_out = at::as_strided(matmul_out, matmul_sizes, strides);
```

Comment:
This feels wrong. IIUC, as_strided creates a view of the input. So if the input tensor is in the wrong memory format there's no way to "relayout" the storage using as_strided. I think you would have caught this problem if [your test](https://github.com/NVIDIA/Fuser/pull/3447/files#diff-22fe4d3f2193276335bba1281b92e245e9ed43e4631a1f17bad63d2cfca0119bR224) had also verified the content (not just the shape) of the output tensor.

Reply:
Agreed! you need an explicit copy here instead.

Reply:
There may be a better way than copy. I suspect matmul_out respects the output strides, but please double check

Reply:
Good call. if we can feed our own outputs there, we might as well do that. So you need to call `as_strided` on output before executing the kernel.

BTW, aten may or may not do that in a fused kernel. i.e. https://github.com/pytorch/pytorch/blob/1f3d8896bc9cea7f46c50ff92b69c6aa139defcb/aten/src/ATen/native/LinearAlgebra.cpp#L2097-L2099

I think the more interesting question is, do we assume it's always a copy and should we try to expose this transpose/copy kernel in fusion for nvfuser to handle it instead... :thinking:

Reply:
> I think the more interesting question is, do we assume it's always a copy and should we try to expose this transpose/copy kernel in fusion for nvfuser to handle it instead...

It's hard to be performance optimal given the implementation of at::matmul_out can change without notifying us. I'd check for the cases that we care about. I suspect for all or 90% of them at::matmul_out is able to fuse. If so, I'd unconditionally assume it'll fuse for simplicity.

Reply:
Ahh you're right. I misunderstood this function call -- there is no copy here. Will fix this to use matmul_out or an explicit copy.